lib/xwax/pitch.h

```diff
@@ -1,71 +1,213 @@
+
 /*
- * Copyright (C) 2021 Mark Hills <mark@xwax.org>
- *
- * This file is part of "xwax".
- *
- * "xwax" is free software; you can redistribute it and/or modify it
- * under the terms of the GNU General Public License, version 3 as
- * published by the Free Software Foundation.
- *
- * "xwax" is distributed in the hope that it will be useful, but
- * WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
- * General Public License for more details.
+ * Kalman filter replacement for alpha-beta pitch estimator
+ * Model: constant velocity (x,v) with acceleration process noise
  *
- * You should have received a copy of the GNU General Public License
- * along with this program; if not, see <https://www.gnu.org/licenses/>.
+ * State:
+ *   x: displacement over the last interval (kept relative by subtracting dx at the end)
+ *   v: velocity (pitch)
  *
+ * Observation:
+ *   z = dx  (measured displacement over dt)
  */
 
 #ifndef PITCH_H
 #define PITCH_H
 
-/* Values for the filter concluded experimentally */
+#include <math.h>
+#include <errno.h>
+#include <stdio.h>
 
-#define ALPHA (1.0/512)
-#define BETA (ALPHA/256)
+struct pitch {
 
-/* State of the pitch calculation filter */
+    /* configuration */
+    double dt;     /* sample period (s) */
+    double *q;      /* process noise spectral density (acceleration), e.g. (rad/s^2)^2 * s */
+    double *r;      /* measurement noise variance for dx (units^2) */
 
-struct pitch {
-    double dt, x, v;
+    /* state */
+    double x;      /* position (relative; we subtract dx each step) */
+    double v;      /* velocity (pitch) */
+
+    /* covariance P (2x2, symmetric) */
+    double Pxx, Pxv, Pvv;
+
+    /* Residual treshold for the mode switch */
+    double mode_threshold;
+
+    /* Stable and reactive parameters for the mode switch */
```

Comment:
Stable is for normal running, reactive for scratching, right? Can you add a comment about it? 
What category are nudging actions to beat match?